[FEATURE] Afficher la nouvelle solution d'une épreuve si présente dans Airtable (PIX-2265).

celineung opened the pull request:
## :unicorn: Problème
On souhaite afficher une réponse personnalisée dans la solution d'une épreuve.

Exemple dans le cas d'une épreuve où:
- les réponse possibles sont 1, 2 ou 3
- on s'attend à avoir un seul chiffre dans l'input de réponse
- on avait dans Airtable le champs "Bonnes réponses" qui contient:
```
1 ou 2 ou 3
1
2
3
```
- dans la correction, on affiche "1 ou 2 ou 3"

Cela pose problème car on mélange deux type d'information dans le champs "Bonnes réponses": les réponses qui sont considérées correctes et celle que l'on veut afficher sur l'écran de solution.

## :robot: Solution
Faite sur https://github.com/1024pix/pix-editor/pull/121:
- créer une nouvelle colonne "Bonnes réponses à afficher" dans la table Epreuves de Airtable
- récupérer ce nouveau champs dans LMCS

Dans cette PR, sur la modale de comparaison des réponses utilisateurs, afficher le texte de la solution à afficher pour tous les types d'épreuves.

## :rainbow: Remarques
Voir https://docs.google.com/presentation/d/1eUK9n5QtX6ijX7KbO-xKDrljmwsOsTbIYUVl6wkI0Ho/edit#slide=id.gc8cf545906_0_59 pour les différents cas.

## :100: Pour tester
- La RA est connectée à Airtable minimal
- Vérifier sur un challenge (ex: recOhUPaWyGdnjN48), que la solution affichée est celle de "Bonnes réponse à afficher" si le champs est rempli

Commit:
Use challenge.solutionToDisplay in correction

pix-service commented:
I'm deploying this PR to these urls:

- App: https://app-pr2703.review.pix.fr
- Orga: https://orga-pr2703.review.pix.fr
- Certif: https://certif-pr2703.review.pix.fr
- Admin: https://admin-pr2703.review.pix.fr
- API: https://api-pr2703.review.pix.fr/api/

 Please check it out!

alexandrecoin commented on api/tests/unit/infrastructure/repositories/correction-repository_test.js:
```diff
@@ -81,25 +72,71 @@ describe('Unit | Repository | correction-repository', function() {
+        challengeDatasource.get.resolves(challengeDataObject);
+
+        // when
+        promise = correctionRepository.getByChallengeId({ challengeId: recordId, userId, locale });
+
+        // then
+        const result = await promise;
```
```suggestion
        const result = await correctionRepository.getByChallengeId({ challengeId: recordId, userId, locale });

        // then
```

**Question** : Y a t-il une raison de ne pas resolve la promesse immédiatement ?

alexandrecoin commented on api/tests/unit/infrastructure/repositories/correction-repository_test.js:
```diff
@@ -81,25 +72,71 @@ describe('Unit | Repository | correction-repository', function() {
+        challengeDatasource.get.resolves(challengeDataObject);
+
+        // when
+        promise = correctionRepository.getByChallengeId({ challengeId: recordId, userId, locale });
+
```
```suggestion
        const result = await correctionRepository.getByChallengeId({ challengeId: recordId, userId, locale });
        
        // then
        return result.hints.forEach((hint) => expect(hint).to.be.an.instanceof(Hint));
```
**Suggestion** : Si pas de raison particulière d'attendre avant de resolve la `promise`

alexandrecoin reviewed (commented):
Tech: OK
Func : En attente du ré-up de la RA. 😊

sbedeau reviewed (commented):
Tech ok
Func en attente de RA

celineung commented:
PR remise en WIP: on souhaite pouvoir afficher une solution dans le cas des embed auto.
Dans ce cas, déplacer la logique du choix de la solution à afficher faite côté api vers le front.

Commit:
Display a challenge's solution

If answer.solutionToDisplay exist, it should alweays be displayed.
Otherwise, if it is not an autoReply challenge, answer.solution is displayed.

MelanieMEB commented on mon-pix/tests/unit/components/comparison-window-test.js:
```diff
@@ -157,4 +157,42 @@ describe('Unit | Component | comparison-window', function() {
+      expect(solution).to.equal('solution');
+    });
+
+    it('should return solutionToDisplay when challenge has autoReply=false and correction has a solutionToDisplay', function() {
```
Il manque le test `should return solutionToDisplay when challenge has autoReply=true and correction has a solutionToDisplay`, et on aura tous les cas :)

celineung replied:
J'ai modifié le nom du test car c'est indépendant du `autoReply`: dès que le `solutionToDisplay` existe, il est affiché.

MelanieMEB commented on mon-pix/app/components/comparison-window.js:
```diff
@@ -68,5 +64,14 @@ export default class ComparisonWindow extends Component {
   }
+
+  get solution() {
+    const solution = this._isAutoReply ? null : this.args.answer.correction.get('solution');
+    return this.args.answer.correction.get('solutionToDisplay') || solution;
```
**question** : est-ce qu'il ne faudrait pas limiter aussi d'après le type de challenge ? 
Je me pose la question de l'impact si on met une `solutionToDisplay` dans un QCU ou dans un QROC, ça risque de casser l'affichage qui se base sur ça ensuite, non ?

celineung replied:
Corrigé pour afficher toujours `solution` dans le cas de QCU et QCM.
Dans les cas de QROC/QROCM:
- si `solutionToDisplay` est présent, il est affiché
- sinon: 
  - si c'est un embedAuto, ne rien afficher
  - sinon afficher `solution`

MelanieMEB reviewed (changes requested):
Globalement OK ! 
Mais je bloque pour qu'on vérifie le cas des qcu/qcm

MelanieMEB committed:
Display by type : Add solutionToDisplay in QCU

MelanieMEB committed:
Display by type : Add solutionToDisplay in QCM

MelanieMEB committed:
Display by type : Add solutionToDisplay in QROC

MelanieMEB committed:
Display by type : Add solutionToDisplay in QROCMDEP

MelanieMEB committed:
Display by type : Add solutionToDisplay in QROCMIND

MelanieMEB committed:
Display by type : Modify comparison window to send solutionToDisplay at each solution panel type

MelanieMEB commented on mon-pix/app/styles/components/_qcm-solution-panel.scss:
```diff
@@ -37,3 +37,7 @@ input.qcm-panel__proposal-checkbox {
+
+.qcm-panel__solution-to-display {
+  margin: 0 0 15px 30px;
+}
```
A supprimer

MelanieMEB commented on mon-pix/app/templates/components/qroc-solution-panel.hbs:
```diff
@@ -1,35 +1,38 @@
   </div>
 </div>
-
+{{/if}}
```
Oubli retour à la ligne

MelanieMEB commented on mon-pix/app/components/qroc-solution-panel.js:
```diff
@@ -27,6 +27,9 @@ export default class QrocSolutionPanel extends Component {
   }
 
   get solutionToDisplay() {
```
Possibilité de renommage : 
`understandableSolution`

MelanieMEB commented on mon-pix/app/templates/components/qroc-solution-panel.hbs:
```diff
@@ -1,35 +1,38 @@
+        {{/if}}
       </div>
+    {{/if}}
+    {{#unless this.isResultOk}}
```
{{#if this.isNotCorrectlyAnswered}} à voir à modifier partout

MelanieMEB commented on mon-pix/app/templates/components/qroc-solution-panel.hbs:
```diff
@@ -1,35 +1,38 @@
+{{#if (or @solution this.solutionToDisplay)}}
```
Sortir dans une méthode : `hasCorrection`

MelanieMEB commented on mon-pix/app/templates/components/qroc-solution-panel.hbs:
```diff
@@ -1,35 +1,38 @@
+{{#if (or @solution this.solutionToDisplay)}}
 <div class="correction-qroc-box rounded-panel">
   <div class="rounded-panel__row ">
```
CSS à modifier ?

MelanieMEB replied:
Inverser le roudned-panel et le correction qrocbox

MelanieMEB commented on mon-pix/app/templates/components/qroc-solution-panel.hbs:
```diff
@@ -1,35 +1,38 @@
+    {{#if @solution}}
+      <div class="correction-qroc-box__answer">
+        {{#if (eq @answer.challenge.format 'paragraphe')}}
+          <textarea class="correction-qroc-box-answer correction-qroc-box-answer--paragraph {{this.inputClass}}"
+                    rows="5"
+                    value="{{this.answerToDisplay}}"
+                    aria-label={{t 'pages.comparison-window.results.a11y.given-answer'}}
+                    disabled>
+          </textarea>
+        {{else if (eq @answer.challenge.format 'phrase')}}
+          <input class="correction-qroc-box-answer correction-qroc-box-answer--sentence {{this.inputClass}}"
+                 value="{{this.answerToDisplay}}"
+                 aria-label={{t 'pages.comparison-window.results.a11y.given-answer'}}
+                 disabled>
+        {{else}}
+          <input class="correction-qroc-box-answer {{this.inputClass}}"
```
Passer par les TextArea et Input ember

MelanieMEB commented on mon-pix/tests/integration/components/qroc-solution-panel-test.js:
```diff
@@ -160,4 +170,129 @@ describe('Integration | Component | QROC solution panel', function() {
+      solution = null;
+    });
+
+    it('should not show the input answer', async function() {
```
doublon avec ci-dessous

MelanieMEB commented on mon-pix/app/components/qrocm-dep-solution-panel.js:
```diff
@@ -41,6 +41,9 @@ export default class QrocmDepSolutionPanel extends Component {
   }
 
   get expectedAnswers() {
```
renommer en understandableSolution

MelanieMEB commented on mon-pix/app/templates/components/qrocm-dep-solution-panel.hbs:
```diff
@@ -27,8 +27,9 @@
       {{/if}}
     {{/each}}
     {{#unless this.answerIsCorrect}}
-      <div class="correction-qrocm__solution">
```
Supprimer le code CSS

MelanieMEB commented on mon-pix/app/templates/components/qrocm-ind-solution-panel.hbs:
```diff
@@ -47,5 +47,16 @@
     {{/each}}
   </div>
+
+  {{#unless this.isResultOk}}
```
Renommer comme vu précédemment

MelanieMEB commented on mon-pix/app/templates/components/qrocm-ind-solution-panel.hbs:
```diff
@@ -47,5 +47,16 @@
   </div>
+
+  {{#unless this.isResultOk}}
+    {{#if @solutionToDisplay}}
```
Pas de test

MelanieMEB commented on mon-pix/app/components/comparison-window.js:
```diff
@@ -66,11 +66,12 @@ export default class ComparisonWindow extends Component {
+    return this._isAutoReply ? null : this.args.answer.correction.get('solution');
+  }
+
+  get solutionToDisplay()
+  {
```
Retour à la ligne

celineung reviewed (commented):
✅ testé fonctionnellement sur les différentes types d'épreuves

alexandrecoin commented:
J'ai ça lors de l'affichage des réponses, un souci de mon côté ? (Testé sur Chrome et Firefox, même résultat)
<img width="987" alt="Capture d’écran 2021-03-22 à 14 28 48" src="https://user-images.githubusercontent.com/36371437/111997190-135a6080-8b1b-11eb-9b5f-99ba4b170cbf.png">
